### PR TITLE
Added API to save image NSData to disk cache.

### PR DESCRIPTION
**Pull request by Mutix**
As discussed in https://github.com/rs/SDWebImage/issues/1065, I thought it would be nice to provide a way to store an image's NSData directly into the disk cache without having to suffer the overhead of creating instances of UIImage.

The new API simply exposes what was the final part of SDImageCache's `storeImage:` function.

It is unit tested and all other tests still pass.

Look forward to any comments.

### CODE REVIEW

**Comment by Mutix** [comment]
@bpoplauschi any thoughts on this?

File: SDWebImage/SDImageCache.m [hunk 1]
```diff
@@ -270,6 +254,27 @@ - (void)storeImage:(UIImage *)image forKey:(NSString *)key toDisk:(BOOL)toDisk {
     [self storeImage:image recalculateFromImage:YES imageData:nil forKey:key toDisk:toDisk];
 }
 
+- (void)storeImageDataToDisk:(NSData *)imageData forKey:(NSString *)key {
+    
+    if (!imageData) return;
```
Comment (bpoplauschi) [inline SDWebImage/SDImageCache.m]: according to our coding style, this should be
```
if (!imageData) {
    return;
}
```

File: Tests/SDWebImage Tests.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -122,6 +122,7 @@
 				DA248D50195472AA00390AB0 /* Frameworks */,
 				DA248D51195472AA00390AB0 /* Resources */,
 				D6347736BDF64FC5A4D078A4 /* Copy Pods Resources */,
+				4B51E412BA3594400947AC71 /* Embed Pods Frameworks */,
```
Comment (bpoplauschi) [inline Tests/SDWebImage Tests.xcodeproj/project.pbxproj]: why this change?

File: Tests/SDWebImage Tests.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -170,6 +171,21 @@
...
+			inputPaths = (
+			);
+			name = "Embed Pods Frameworks";
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "\"${SRCROOT}/Pods/Target Support Files/Pods-ios/Pods-ios-frameworks.sh\"\n";
+			showEnvVarsInLog = 0;
+		};
```
Comment (bpoplauschi) [inline Tests/SDWebImage Tests.xcodeproj/project.pbxproj]: why adding this shell script?

**Comment by bpoplauschi** [comment]
Looks good, aside from the small comments I added.

File: SDWebImage/SDImageCache.m (hunk 1 above)
Comment (Mutix) [inline SDWebImage/SDImageCache.m]: Ok no problem, I'll align to your coding style. Is there a doc somewhere where the style to adhere to is defined?

File: Tests/SDWebImage Tests.xcodeproj/project.pbxproj (hunk 2 above)
Comment (Mutix) [inline Tests/SDWebImage Tests.xcodeproj/project.pbxproj]: I imagine the change might be due to the version of Cocoapods I used to build the test project. I believe it was 0.39.0. Cocoapods adds these build phases when running 'pod install'

File: Tests/SDWebImage Tests.xcodeproj/project.pbxproj (hunk 3 above)
Comment (Mutix) [inline Tests/SDWebImage Tests.xcodeproj/project.pbxproj]: See line comment above - this is a build phase generated by Cocoapods pod install

File: Tests/Tests/SDImageCacheTests.m [hunk 4]
```diff
@@ -193,9 +212,14 @@ - (void)clearAllCaches{
...
+    return [UIImage imageWithContentsOfFile:[self testImagePath]];
+}
+
+- (NSString *)testImagePath {
+    
+    NSBundle *testBundle = [NSBundle bundleForClass:[self class]];
+    return [testBundle pathForResource:@"TestImage" ofType:@"jpg"];
 }
 
 @end
```
Comment (Mutix) [inline Tests/Tests/SDImageCacheTests.m]: I'll also make sure there's a new line here!

Comment (Mutix) [inline Tests/Tests/SDImageCacheTests.m]: I'll also make sure there's a new line here

**Comment by fabiosoft** [comment]
How can I retrieve the stored NSData? Now only methods returning UIImage are available

**Comment by bpoplauschi** [comment]
@fabiosoft do you mean from `SDImageCache`? Depends on the SDWebImage version you are using. For versions 4.x, you can use `diskImageDataForKey:`

**Comment by fabiosoft** [comment]
I'm using  v4.3.3, I need to save a downloaded NSData (it's a gif) and use a memory cache (no persisting on disk needed)

using 
```
[[SDImageCache sharedImageCache]storeImage:nil imageData:data forKey:item.identifier toDisk:NO completion:^{}];
```

I hope it's stored, but how can I retrieve the NSData saved?
I can only see methods returning UIImages

```
- (nullable UIImage *)imageFromMemoryCacheForKey:(nullable NSString *)key;
- (nullable UIImage *)imageFromDiskCacheForKey:(nullable NSString *)key;
- (nullable UIImage *)imageFromCacheForKey:(nullable NSString *)key;
```

**Comment by dreampiggy** [comment]
This API is available in v4.4.2. See https://github.com/rs/SDWebImage/releases/tag/4.4.2

**Comment by fabiosoft** [comment]
Cool, but it seems to use disk cache only... is possible to use memory cache?

**Comment by dreampiggy** [comment]
@fabiosoft `SDMemoryCache` only store the `UIImage` instance in the memory, no any image data is kept in the memory. So you can not get any image data in the memory.

If you need to save the original image data in the memory , consider to use [Custom Cache](https://github.com/rs/SDWebImage/wiki/Advanced-Usage#cache-protocol) feature in the 5.x to write your own. You may need to implement the protocol method `storeImage:imageData:cacheType:completion:` to write image data to the memory cache as well.
